ui/pages/asset/components/asset-chart.tsx

```diff
@@ -0,0 +1,349 @@
+
+    fetchWithCache({
+      url: `https://price-api.metafi.codefi.network/v1/chains/${chainId}/historical-prices/${address}?vsCurrency=${currency}&timePeriod=${timeRange}`,
+      cacheOptions: { cacheRefreshTime: MINUTE },
+      functionName: 'GetAssetHistoricalPrices',
+    })
+      .catch(() => ({}))
+      .then((resp?: { prices?: number[][] }) => {
+        const prices = resp?.prices?.map((p) => ({ x: p?.[0], y: p?.[1] }));
+
+        if (prices && prices?.length > 0) {
+          priceRef?.current?.setPrices({
+            price: undefined,
+            comparePrice: prices[0].y,
+            date: Date.now(),
+          });
+
+          let [xMin, xMax, yMin, yMax]: Point[] = [];
+          for (const p of prices) {
+            xMin = !xMin || p.x < xMin.x ? p : xMin;
+            xMax = !xMax || p.x > xMax.x ? p : xMax;
+            yMin = !yMin || p.y < yMin.y ? p : yMin;
+            yMax = !yMax || p.y > yMax.y ? p : yMax;
+          }
+
+          const drawTooltips = () => {
+            maxPriceTooltip.current?.setTooltip({
+              xAxisPercent: (yMax.x - xMin.x) / (xMax.x - xMin.x),
+              price: yMax.y,
+            });
+            minPriceTooltip.current?.setTooltip({
+              xAxisPercent: (yMin.x - xMin.x) / (xMax.x - xMin.x),
+              price: yMin.y,
+            });
+          };
+
+          setChartOptions((options) => ({
+            ...options,
+            onResize: () => drawTooltips(),
+            scales: {
+              x: { min: xMin.x, max: xMax.x, display: false, type: 'linear' },
+              y: { min: yMin.y, max: yMax.y, display: false },
+            },
+          }));
+          setData({ datasets: [{ data: prices }] });
+          drawTooltips();
+        }
+
+        setLoading(false);
+      });
+  }, [chainId, address, currency, timeRange]);
+
+  useEffect(
+    () =>
+      setChartOptions((options) => ({
+        ...options,
+        borderColor: theme === 'dark' ? brandColor.blue400 : brandColor.blue500,
+      })),
+    [theme],
+  );
+
+  const prices = chartRef?.current?.data?.datasets?.[0]?.data;
+
+  return (
+    <Box>
+      <AssetPrice
+        currentPrice={currentPrice}
+        loading={loading}
+        ref={priceRef}
+      />
+      <Box
+        style={{ opacity: prices && loading ? 0.2 : 1 }}
+        backgroundColor={
+          loading && !prices
+            ? BackgroundColor.backgroundAlternative
+            : BackgroundColor.backgroundDefault
+        }
+      >
+        <ChartTooltip ref={maxPriceTooltip} />
+        <Box
+          display={Display.Flex}
+          justifyContent={JustifyContent.center}
+          style={{
+            aspectRatio: `${chartOptions.aspectRatio}`,
+          }}
+        >
+          {data ? (
+            <Line
+              ref={chartRef}
+              data={data}
+              options={chartOptions}
+              updateMode="none"
+              onMouseMove={({ nativeEvent: e }: ReactMouseEvent) => {
```

Comment:
you can use `useCallback` to handle event, this will avoid unnecessary re-renders and optimize perf

Reply:
Are you suggesting a "debounce" kind of optimization, to prevent it from running too often?